pypeit/edgetrace.py

```diff
@@ -3948,61 +3965,80 @@ def sync(self, rebuild_pca=True, debug=False):
-                      'have been placed close to the detector edges. Flagging '
-                      '{0} traces that are not properly sorted for removal.'.format(np.sum(indx)))
-            # Mask the traces as due to a synchronization error
-            # NOTE: These are only masked here so that they can be
-            # plotted if debug is True. Because the full traces are
-            # masked, they're immediately removed by check_synced.
-            self.edge_msk[:,indx] = self.bitmask.turn_on(self.edge_msk[:,indx], 'SYNCERROR')
-
-        if debug:
-            msgs.info('Show instance includes inserted traces but before checking the sync.')
-            self.show(flag='any')
+        # NOTE: The maximum number of iterations is hard-coded for now.  Testing
+        # is needed to know if we need this to be any larger than 3.
+        maxiter = 3
+        i = 0
+        while i < maxiter:
+            msgs.info(f'Beginning syncing iteration : {i+1} (of at most {maxiter})')
+
+            # Find the edges to add, what side they're on, and where to insert
+            # them into the existing trace array.  This is done again, in case
+            # we're going through another iteration
+            side, add_edge, add_indx = self._get_insert_locations()
+
+            # Get the reference locations for the new edges
+            trace_ref = self._get_reference_locations(trace_cen, add_edge)
+
+            # Determine which sync_predict to use
+            if self.par['sync_predict'] == 'pca' \
+                    or (self.par['sync_predict'] == 'auto' and self.can_pca()):
+                _sync_predict = 'pca'
+            else:
+                _sync_predict = 'nearest'
+
+            # Predict the traces either using the PCA or using the nearest slit edge
+            if _sync_predict == 'pca':
+                trace_add = self.predict_traces(trace_ref[add_edge], side=side[add_edge])
+            elif _sync_predict == 'nearest':
+                # Index of trace nearest the ones to add
+                # TODO: Force it to use the nearest edge of the same side;
+                # i.e., when inserting a new right, force it to use the
+                # nearest right instead of the nearest left?
+                nearest = utils.nearest_unmasked(np.ma.MaskedArray(trace_ref, mask=add_edge))
+                # Indices of the original traces
+                indx = np.zeros(len(add_edge), dtype=int)
+                indx[np.logical_not(add_edge)] = np.arange(self.ntrace)
+                # Offset the original traces by a constant based on the
+                # reference trace position to construct the new traces.
+                trace_add = trace_cen[:,indx[nearest[add_edge]]] + trace_ref[add_edge] \
+                                - trace_ref[nearest[add_edge]]
+
+            # Insert the new traces and resort them spatially
+            self.insert_traces(side[add_edge], trace_add, loc=add_indx[add_edge], mode='sync')
```

Comment:
after this `trace_cen` needs to be updated, otherwise in the following iteration there can be an index failure

Reply:
Good catch!  Fixed.